app/models/foreman_azure_rm/azure_rm.rb

```diff
@@ -187,55 +203,56 @@ def user_data_supported?
+          args[:script_command] =  sudoers_cmd
         end
         disable_password_auth = false
-      elsif args[:azure_vm][:ssh_key_data].present? && !args[:azure_vm][:password].present?
+      elsif args[:ssh_key_data].present? && !args[:password].present?
         disable_password_auth = true
       else
         disable_password_auth = false
       end
+
       vm             = create_managed_virtual_machine(
-        name:                            args[:azure_vm][:vm_name],
-        location:                        args[:azure_vm][:location],
-        resource_group:                  args[:azure_vm][:resource_group],
-        vm_size:                         args[:azure_vm][:vm_size],
-        username:                        args[:azure_vm][:username],
-        password:                        args[:azure_vm][:password],
-        ssh_key_data:                    args[:azure_vm][:ssh_key_data],
+        name:                            args[:vm_name],
+        location:                        region,
+        resource_group:                  args[:resource_group],
+        vm_size:                         args[:vm_size],
+        username:                        args[:username],
+        password:                        args[:password],
+        ssh_key_data:                    args[:ssh_key_data],
         disable_password_authentication: disable_password_auth,
         network_interface_card_ids:      nics.map(&:id),
-        platform:                        args[:azure_vm][:platform],
+        platform:                        args[:platform],
         vhd_path:                        args[:image_id],
-        os_disk_caching:                 args[:azure_vm][:os_disk_caching],
-        premium_os_disk:                 args[:azure_vm][:premium_os_disk],
+        os_disk_caching:                 args[:os_disk_caching],
+        premium_os_disk:                 args[:premium_os_disk],
         custom_data:                     args[:user_data],
-        script_command:                  args[:azure_vm][:script_command],
-        script_uris:                     args[:azure_vm][:script_uris],
+        script_command:                  args[:script_command],
+        script_uris:                     args[:script_uris],
       )
-      create_vm_extension(args)
+      logger.debug "Virtual Machine #{args[:vm_name]} Created Successfully."
+      create_vm_extension(region, args)
       # return the vm object using azure_vm
-      return_vm = AzureRMCompute.new(azure_vm: vm, sdk: sdk)
+      return_vm = AzureRMCompute.new(azure_vm: vm, sdk: sdk, resource_group: args[:resource_group], nics: vm_nics(vm))
     rescue RuntimeError => e
       Foreman::Logging.exception('Unhandled Azure RM error', e)
       destroy_vm vm.id if vm
       raise e
     end
 
     def destroy_vm(uuid)
-      #vm.azure_vm because that's the azure object and vm is the wrapper
       vm           = find_vm_by_uuid(uuid)
       vm_name      = vm.name
```

Comment:
Why not to use `vm.name` directly as used only once? Similar case for `nic_ids`.